components/messenger.rst

```diff
@@ -86,6 +86,65 @@ that will do the required processing for your message::
        }
     }
 
+Envelope
+--------
+
+The notion of an envelope is a concept that helps add context around the
+messages. An envelope is a message and a set of data. From a user's perspective, this
+allows you to set some configuration around the message. For example, to set the serialization
+groups used when the message goes through the transport layer, wrap your message
+in an ``Envelope`` and add some ``SerializerConfiguration``::
+
+    use Symfony\Component\Messenger\Envelope;
+    use Symfony\Component\Messenger\Transport\Serialization\SerializerConfiguration;
+
+    $bus->dispatch(
+        (new Envelope($message))->with(new SerializerConfiguration([
+            'groups' => ['my_serialization_groups'],
+        ]))
+    );
+
+At the moment, the Symfony Messenger has the following built-in envelopes:
+
+1. :class:`Symfony\\Component\\Messenger\\Transport\\Serialization\\SerializerConfiguration`,
+   to configure the serialization groups used by the transport.
+2. :class:`Symfony\\Component\\Messenger\\Middleware\\Configuration\\ValidationConfiguration`,
+   to configure the validation groups used when the validation middleware is enabled.
+3. :class:`Symfony\\Component\\Messenger\\Asynchronous\\Transport\\ReceivedMessage`,
+   an internal item that marks the message as received from a transport.
+
+Instead of dealing directly with the messages in the middleware you can receive the
+envelope by implementing the :class:`Symfony\\Component\\Messenger\\EnvelopeAwareInterface`
+marker, like this::
+
+    use Symfony\Component\Messenger\Asynchronous\Transport\ReceivedMessage;
+    use Symfony\Component\Messenger\Middleware\MiddlewareInterface;
+    use Symfony\Component\Messenger\EnvelopeAwareInterface;
+
+    class MyOwnMiddleware implements MiddlewareInterface, EnvelopeAwareInterface
+    {
+        public function handle($message, callable $next)
+        {
+            // $message here is an `Envelope` object, because this middleware
+            // implements the EnvelopeAwareInterface interface. Otherwise,
+            // it would be the "original" message.
+
+            if (null !== $message->get(ReceivedMessage::class)) {
```

Comment:
Is this `get()` method something that lives on `Envelope`? Isn't there always just one message inside an Envelope? If so, why do we pass an argument here? Or, could you also pass something like `$message->get(SerializerConfiguration::class)` to get those types of things out?

Also, suppose some similar code: if `$message->get(MyMessage::class)` is null, does it mean that this message is NOT originally a `MyMessage` isntance, right? I mean, it's equivalent to `!$message instanceof MyMessage` in a middleware without `EnvelopAwareInterface`?

Some inline comments with some of these answers might be all we need :)

Reply:
Also, is it a real use-case to look need to know whether or not a message was just received? If you're using an async transport and you want to do something for a specific message type, is it important that you make sure this processing is done only when your message has been "received"? If so, why? And if so, how could I both perform my action only when the message is "received" AND only check the class of the original message (e.g. I need to do something when my `MyMessage` class is "received"? Or is that totally not the correct pattern.

Reply:
> if $message->get(MyMessage::class) is null, does it mean that this message is NOT originally a MyMessage isntance, right?

There is mis interpretation here. The `ReceivedMessage` class is not an example here, it's an actual envelope item within the component has is a marker that represents that the message has just been received. It's needed to have the router middleware ignoring them (otherwise we would have a receive -> send) loop.

> Or, could you also pass something like $message->get(SerializerConfiguration::class) to get those types of things out?

That's not _also_, that's the only thing you can do :)

> how could I both perform my action only when the message is "received" AND only check the class of the original message

```php
if (nul !== $message->get(ReceivedMessage::class) && $message->getMessage() instanceof YourOwnMessageClass) {
    // Your logic...
}
```

Might be clearer if we [rename `$message` to `$envelope`](#)?